### PR TITLE
Testable Loopia API Struct

### PR DESCRIPTION
**Pull request by jonlil**
@akarl818 review? :- )

### CODE REVIEW

File: loopia.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  "github.com/kolo/xmlrpc"
+)
+
+const API_URL string = "https://api.loopia.se/RPCSERV"
+
+type API struct {
+    APIUsername string
+    APIPassword string
+    RCPEndpoint string
+    xmlrpc *xmlrpc.Client
```
Comment (jonlil) [inline loopia.go]: woops

File: loopia_test.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+func teardown() {
+    server.Close()
+}
+
+func TestClient_Credentials(t *testing.T) {
+    setup()
+    mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+        assert.Equal(t, "POST", r.Method, "Expected method 'POST', got %s", r.Method)
```
Comment (akarl) [inline loopia_test.go]: Are you sure this assert can fail properly?

File: loopia_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        body, err := ioutil.ReadAll(r.Body)
+        if err != nil {
+            fmt.Println(err)
```
Comment (akarl) [inline loopia_test.go]: Should this be a test failure?

File: loopia.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  return api, nil
+}
+
+func (api *API) XmlRpcClient() *xmlrpc.Client {
+  client, _ := xmlrpc.NewClient(api.RCPEndpoint, nil)
+  return client
+}
+
+func New(username string, password string) (*API, error) {
```
Comment (akarl) [inline loopia.go:24]: Is there a reason `New` doesn't just `return &API{...}`?

File: dns.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    if err := api.XmlRpcClient().Call("getZoneRecords", args, &result); err != nil {
+      return []Record{}, err
+    }
+
+    return result, nil
+}
+
+func (api *API) GetZoneRecord(domain string, subdomain string, id int64) (Record, error) {
```
Comment (akarl) [inline dns.go]: Is there a reason why these functions don't return pointers?

File: dns.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package loopia
+
+import (
+    "errors"
+)
+
+type Record struct {
+    Id int64 `xmlrpc:"record_id"`
+    Ttl int `xmlrpc:"ttl"`
```
Comment (akarl) [inline dns.go]: Ttl should be TTL, and Id should be ID (Go conventions) :)

File: loopia_test.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    mux *http.ServeMux
+
+    // client is the API client being tested
+    client *API
+
+    // server is a test HTTP server used to provide mock API responses
+    server *httptest.Server
+)
+
+func setup() {
```
Comment (akarl) [inline loopia_test.go:23]: If you don't want to use global variables you can make this function take a `http.HandlerFunc` as an argument. Then pass that to `httptest.NewServer` and skip the mux.

**Review by akarl — APPROVED** [review]
Nice!

Don't forget to `go lint`, `go fmt` and documentation.

File: dns.go (hunk 5 above)
Comment (jonlil) [reply dns.go]: Not really, I need to read up a lot on pointers :-)

File: loopia_test.go (hunk 7 above)
Comment (jonlil) [reply loopia_test.go:23]: And instead call `server.HandleFunc`?

File: loopia.go (hunk 4 above)
Comment (jonlil) [reply loopia.go:24]: Not really, refactored :-) Thanks for your comment

**Pull request merged**